### PR TITLE
r/aws_s3_bucket: support SSE-KMS replication configuration

### LINKED ISSUE #2200: Feature request: S3 Cross-Region Replication with KMS
Hi,

Just a couple of hours ago, AWS introduced S3 CRR with KMS support. It would nice to get this integrated into terraform (though it seems that support for it is lacking in AWS Go SDK)

https://aws.amazon.com/blogs/aws/new-amazon-s3-encryption-security-features/

### PR DESCRIPTION
**Pull request by modax**
Fixes #2200, #2226

It would be great to get this merged soon. Thanks!

```
make testacc TESTARGS="-run 'TestAccAWSS3Bucket_'"                                                   
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run 'TestAccAWSS3Bucket_' -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSS3Bucket_importBasic
--- PASS: TestAccAWSS3Bucket_importBasic (79.48s)
=== RUN   TestAccAWSS3Bucket_importWithPolicy
--- PASS: TestAccAWSS3Bucket_importWithPolicy (86.50s)
=== RUN   TestAccAWSS3Bucket_basic
--- PASS: TestAccAWSS3Bucket_basic (65.44s)
=== RUN   TestAccAWSS3Bucket_namePrefix
--- PASS: TestAccAWSS3Bucket_namePrefix (65.88s)
=== RUN   TestAccAWSS3Bucket_generatedName
--- PASS: TestAccAWSS3Bucket_generatedName (66.26s)
=== RUN   TestAccAWSS3Bucket_region
--- PASS: TestAccAWSS3Bucket_region (39.46s)
=== RUN   TestAccAWSS3Bucket_acceleration
--- PASS: TestAccAWSS3Bucket_acceleration (73.89s)
=== RUN   TestAccAWSS3Bucket_RequestPayer
--- PASS: TestAccAWSS3Bucket_RequestPayer (123.23s)
=== RUN   TestAccAWSS3Bucket_Policy
--- PASS: TestAccAWSS3Bucket_Policy (168.01s)
=== RUN   TestAccAWSS3Bucket_UpdateAcl
--- PASS: TestAccAWSS3Bucket_UpdateAcl (120.25s)
=== RUN   TestAccAWSS3Bucket_Website_Simple
--- PASS: TestAccAWSS3Bucket_Website_Simple (183.89s)
=== RUN   TestAccAWSS3Bucket_WebsiteRedirect
--- PASS: TestAccAWSS3Bucket_WebsiteRedirect (185.42s)
=== RUN   TestAccAWSS3Bucket_WebsiteRoutingRules
--- PASS: TestAccAWSS3Bucket_WebsiteRoutingRules (128.49s)
=== RUN   TestAccAWSS3Bucket_enableDefaultEncryption_whenTypical
--- PASS: TestAccAWSS3Bucket_enableDefaultEncryption_whenTypical (109.67s)
=== RUN   TestAccAWSS3Bucket_enableDefaultEncryption_whenAES256IsUsed
--- PASS: TestAccAWSS3Bucket_enableDefaultEncryption_whenAES256IsUsed (67.26s)
=== RUN   TestAccAWSS3Bucket_disableDefaultEncryption_whenDefaultEncryptionIsEnabled
--- PASS: TestAccAWSS3Bucket_disableDefaultEncryption_whenDefaultEncryptionIsEnabled (124.96s)
=== RUN   TestAccAWSS3Bucket_shouldFailNotFound
--- PASS: TestAccAWSS3Bucket_shouldFailNotFound (36.49s)
=== RUN   TestAccAWSS3Bucket_Versioning
--- PASS: TestAccAWSS3Bucket_Versioning (178.60s)
=== RUN   TestAccAWSS3Bucket_Cors
--- PASS: TestAccAWSS3Bucket_Cors (131.24s)
=== RUN   TestAccAWSS3Bucket_Logging
--- PASS: TestAccAWSS3Bucket_Logging (103.32s)
=== RUN   TestAccAWSS3Bucket_Lifecycle
--- PASS: TestAccAWSS3Bucket_Lifecycle (176.16s)
=== RUN   TestAccAWSS3Bucket_Replication
--- PASS: TestAccAWSS3Bucket_Replication (254.52s)
=== RUN   TestAccAWSS3Bucket_ReplicationWithoutStorageClass
--- PASS: TestAccAWSS3Bucket_ReplicationWithoutStorageClass (98.22s)
=== RUN   TestAccAWSS3Bucket_ReplicationExpectVersioningValidationError
--- PASS: TestAccAWSS3Bucket_ReplicationExpectVersioningValidationError (50.31s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       2716.962s
```

### CODE REVIEW

**Comment by modax** [comment]
Glad to see a release with #2472. This one would have made a nice pairing (and it still does)!

**Comment by cornfeedhobo** [comment]
@radeksimko can you urge a review of this? I'm pretty excited to see this merged.

**Comment by cornfeedhobo** [comment]
@jen20 Maybe this PR would be suitable for your review?

**Comment by cornfeedhobo** [comment]
Bump. This PR is fairly trivial to review and has been ready for nearly a month. Please review.

**Comment by cornfeedhobo** [comment]
@radeksimko could this get a review? It's blocking a bunch of work for me here.

**Comment by johvet** [comment]
We'd also love to see this coming soon.

**Comment by cornfeedhobo** [comment]
@jen20 @radeksimko any hope here?

**Comment by cornfeedhobo** [comment]
can haz merge?

**Comment by cornfeedhobo** [comment]
@modax Looks like you might have to take another pass here. Thanks for owning this!

**Comment by jrstarke** [comment]
@modax: looks like this has conflicts now

**Comment by jrstarke** [comment]
Looks like there are some conflicts. If we get the conflicts resolved, can we get this merged?

**Comment by modax** [comment]
Fixed conflicts (removed two commits addressing other related problems that already been fixed in master), reran acc tests.

However, it is really unfortunate that PR merging cycle takes such a long time...

**Comment by tombuildsstuff** [comment]
Tests pass:

screen shot 2018-02-19 at 18 37 57

**Comment by modax** [comment]
@tombuildsstuff Thanks for review. Sorry for the late change, but I've just realized that my `testAccCheckAWSS3BucketReplicationRules` was still using old "provider for loop" which deemed to be deprecated and was refactored in `testAccCheckAWSS3BucketExistsWithProvider` with specific provider function. So I have refactored my func in the spirit of the new  `testAccCheckAWSS3BucketExistsWithProvider` as well: https://github.com/terraform-providers/terraform-provider-aws/pull/2625/commits/2b643af8e8223a9cfa301f53d91158c176e22394

Hopefully this does not delay merge...

```
$ make testacc TESTARGS="-run 'TestAccAWSS3Bucket_Replication'" 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run 'TestAccAWSS3Bucket_Replication' -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSS3Bucket_Replication
--- PASS: TestAccAWSS3Bucket_Replication (271.27s)
=== RUN   TestAccAWSS3Bucket_ReplicationWithoutStorageClass
--- PASS: TestAccAWSS3Bucket_ReplicationWithoutStorageClass (104.66s)
=== RUN   TestAccAWSS3Bucket_ReplicationExpectVersioningValidationError
--- PASS: TestAccAWSS3Bucket_ReplicationExpectVersioningValidationError (54.79s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       430.744s
```

File: aws/resource_aws_s3_bucket.go [hunk 1]
```diff
@@ -2086,6 +2154,12 @@ func rulesHash(v interface{}) int {
 	if v, ok := m["status"]; ok {
 		buf.WriteString(fmt.Sprintf("%s-", v.(string)))
 	}
+	if v, ok := m["destination"].(*schema.Set); ok && v.Len() > 0 {
+		buf.WriteString(fmt.Sprintf("%d-", destinationHash(v.List()[0])))
+	}
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply aws/resource_aws_s3_bucket.go]: In that case I think we should add state migration, so the diff doesn't appear for existing users who just upgraded and not use any new functionality.

[the remark this answers is not in the document]
Comment (modax) [reply aws/resource_aws_s3_bucket.go]: hi, @radeksimko,

I have just confirmed that neither of these:

1) terraform apply
2) terraform apply -refresh=false
3) terraform refresh

report any changes after provider upgrade. Rule hash value changes in the state file after all of them though. Frankly, I have also expected 2) to report changes but for some reason it does not happen (must be related to how set hash is compared probably, i.e. always upfront or something).

[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply aws/resource_aws_s3_bucket.go]: confirmed that scenario ^ - LGTM 👍

**Comment by bflad** [comment]
This has been released in [version 1.10.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1100-february-24-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by cornfeedhobo** [comment]
This seriously might make my quarter. Thank you @modax for getting this in!!

**Comment by jrstarke** [comment]
Oh man, this has me super bummed. I thought that this was exactly what I needed, but it turns out I also need the `Account` and `AccessControlTranslation` arguments. Without those, all of the files getting replicated to my replication account are un-usable, unless the owner account explicitly grants access to all of them.

**Comment by jnoss** [comment]
+1 for #3575 to get this working for cross-account replication, too

**Pull request merged**